File: src/bin/sample.rs
use oxilearn::{
    cart_pole::CartPole,
    dqn::DoubleDeepAgent,
    epsilon_greedy::EpsilonGreedy,
    experience_buffer::RandomExperienceBuffer,
    generate_policy,
    trainer::{TrainResults, Trainer},
    OxiLearnErr,
};
use std::env;
use tch::{Device, Tensor};
fn main() {
    let args: Vec<String> = env::args().collect();
    let seed = args[1].parse::<u64>().unwrap();
    tch::manual_seed(seed as i64);
    tch::maybe_init_cuda();

    let device = Device::Cpu;
    let train_env = CartPole::new(500, seed);
    let eval_env = CartPole::new(500, seed + 1);

    let update_strategy =
        oxilearn::epsilon_greedy::EpsilonUpdateStrategy::EpsilonLinearTrainingDecreasing {
            start: 1.0,
            end: 0.05,
            end_fraction: 0.1,
        };
    let action_selector = EpsilonGreedy::new(1.0, seed + 2, update_strategy);

    let mem_replay = RandomExperienceBuffer::new(10_000, 4, 1_000, seed + 3, false, device);
    let policy = generate_policy(
        vec![
            (256, |xs: &Tensor| xs.relu()),
            (256, |xs: &Tensor| xs.relu()),
        ],
        |xs: &Tensor| xs.shallow_clone(),
        4,
        2,
    )
    .unwrap();
    let opt = oxilearn::dqn::OptimizerEnum::Adam(tch::nn::Adam::default());
    let loss_fn = |pred: &Tensor, target: &Tensor| pred.mse_loss(target, tch::Reduction::Mean);

    let mut model = DoubleDeepAgent::new(
        action_selector,
        mem_replay,
        policy,
        opt,
        loss_fn,
        0.03,
        0.99,
        10.0,
        device,
    );

    let mut trainer = Trainer::new(train_env, eval_env);
    trainer.early_stop = Some(Box::new(move |reward| reward >= 475.0));

    let training_results: Result<TrainResults, OxiLearnErr> =
        trainer.train_by_steps(&mut model, 100_000, 128, 256, 256, 10, 1000, 10, 1);
    let evaluation_results = trainer.evaluate(&mut model, 10);
    println!("Evaluation results: {:?}", evaluation_results);
}

File: python-src/sample.py
import gymnasium as gym
import sys
import torch
from torch import nn
from dqn import DoubleDeepAgent, OptimizerEnum, generate_policy
from epsilon_greedy import EpsilonGreedy, EpsilonUpdateStrategy
from experience_buffer import RandomExperienceBuffer
from trainer import Trainer, TrainResults


def main():
    args = sys.argv
    seed = int(args[1])
    torch.manual_seed(seed)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    train_env = gym.make("CartPole-v1")
    train_env.reset(seed=seed)
    eval_env = gym.make("CartPole-v1")
    eval_env.reset(seed=seed + 1)

    update_strategy = EpsilonUpdateStrategy.EpsilonLinearTrainingDecreasing(
        start=1.0, end=0.04, end_fraction=0.16
    )
    action_selector = EpsilonGreedy(1.0, seed + 2, update_strategy)

    mem_replay = RandomExperienceBuffer(100_000, 4, 1_000, seed + 3, False, device)

    policy = generate_policy(
        [
            (256, nn.ReLU()),
            (256, nn.ReLU()),
        ],
        nn.Softmax(),
        4,
        2,
    )

    optimizer = torch.optim.Adam
    loss_fn = nn.MSELoss()

    model = DoubleDeepAgent(
        action_selector,
        mem_replay,
        policy,
        optimizer,
        loss_fn,
        2.3e-3,
        0.99,
        10.0,
        device,
    )

    trainer = Trainer(train_env, eval_env)
    trainer.early_stop = lambda reward: reward >= 475.0

    training_results = trainer.train_by_steps(
        model, 50_000, 128, 256, 64, 10, 1000, 10, 1
    )
    training_steps = sum(training_results[1])

    evaluation_results = trainer.evaluate(model, 10)
    rewards = evaluation_results[0]
    reward_avg = sum(rewards) / len(rewards)
    variance = sum((reward_avg - value) ** 2 for value in rewards) / len(rewards)
    std = variance**0.5

    print(f"python,{seed},{training_steps},{reward_avg},{std}")


if __name__ == "__main__":
    main()
